scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -511,3 +513,391 @@ def _adjoint(self):
+
+    >>> sak.toarray()
+    array([[ 5, -4,  1,  0,  0,  0],
+           [-4,  6, -4,  1,  0,  0],
+           [ 1, -4,  6, -4,  1,  0],
+           [ 0,  1, -4,  6, -4,  1],
+           [ 0,  0,  1, -4,  6, -4],
+           [ 0,  0,  0,  1, -4,  5]])
+    >>> sak.tobanded()
+    array([[ 1,  1,  1,  1,  1,  1],
+           [-4, -4, -4, -4, -4, -4],
+           [ 5,  6,  6,  6,  6,  5]])
+    >>> sak.tosparse()
+    <6x6 sparse matrix of type '<class 'numpy.int32'>'
+        with 24 stored elements (5 diagonals) in DIAgonal format>
+    >>> np.array_equal(sak.tosparse().toarray(), sak.toarray())
+    True
+    >>> np.array_equal(sak(np.eye(n)), sak.tosparse().toarray())
+    True
+    >>> sak.eigenvalues
+    array([0.03922866, 0.56703972, 2.41789479, 5.97822974,
+           10.54287655, 14.45473055])
+
+    The banded form can be used in scipy functions for banded matrices, e.g.,
+
+    >>> e = eig_banded(sak.tobanded(), eigvals_only=True)
+    >>> np.allclose(sak.eigenvalues, e, atol= n * n * n * np.finfo(float).eps)
+    True
+
+    """
+    def __init__(self, n, dtype=np.int32):
+        self.n = n
+        self.dtype = dtype
+        shape = (n, n)
+        super().__init__(dtype, shape)
+
+        k = np.arange(1, n + 1)
+        e = np.sort(16. * np.power(np.cos(0.5 * k * np.pi / (n + 1)), 4))
+        self._eigenvalues = e
+
+    @property
+    def eigenvalues(self):
+        return self._eigenvalues
+
+    @eigenvalues.setter
+    def eigenvalues(self, value):
+        raise AttributeError('"eigenvalues" attribute is read-only and cannot '
+                             'be set.')
+
+    def tosparse(self):
+        from scipy.sparse import spdiags
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n, dtype=self.dtype)
+        d2 = np.ones(self.n, dtype=self.dtype)
+        return spdiags([d2, d1, d0, d1, d2], [-2, -1, 0, 1, 2], self.n, self.n)
+
+    def tobanded(self):
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n, dtype=self.dtype)
+        d2 = np.ones(self.n, dtype=self.dtype)
+        return np.array([d2, d1, d0])
+
+    def toarray(self):
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n - 1, dtype=self.dtype)
+        d2 = np.ones(self.n - 2, dtype=self.dtype)
+        a = np.diag(d0)
+        a += np.diag(d1, 1) + np.diag(d1, -1)
+        a += np.diag(d2, 2) + np.diag(d2, -2)
+        return a
+    
+    def _matvec(self, x):
+        n = self.n
+        assert n == x.shape[0]
+        x = x.reshape(n, -1)
+        sx = np.zeros_like(x)
+        sx[0,:] = 5 * x[0,:] - 4 * x[1,:] + x[2,:]
+        sx[-1,:] = 5 * x[-1,:] - 4 * x[-2,:] + x[-3,:]
+        sx[1:-1,:] = (6 * x[1:-1,:] - 4 * (x[:-2,:] + x[2:,:])
+                      + np.pad(x[:-3,:], ((1,0),(0,0)))
+                      + np.pad(x[3:,:], ((0,1),(0,0))))
```

Comment:
Upon further inspection, this seems like just a manual application of the matrix-vector multiplication rules, together with the known banded structure?

Why don't we just do `self.toarray() @ x`? Or, if performance is an issue, `self.tosparse() @ x` and _then_ materialise the vector?

Reply:
> Because the whole point of the `LinearOperator.matvec()` infrastructure is to avoid explicitly constructing either of these.

I can see the point of not constructing the full matrix. But I'd be interested how much slower things would be with the sparse matrix multiplication (I'd be surprised if it's much worse in either runtime or memory usage). It just seems like an unnecessary source of errors to me to recode this by hand.

Reply:
If there is an error, it's easily found out during the comparisons between all of the different paths. Because this is all low-tier testing/benchmarking code, this isn't going to see high traffic for modifications, so if bugs aren't in there now, I don't think they'll be introduced later.